### DISCUSSION TITLE
Turbo module / Fabric support (New Architecture)?

### DISCUSSION THREAD

**Discussion by mfbx9da4** [invertase/react-native-firebase#6200] [Ideas] [answered]
Is RN fire base planning to move to turbo module initialisation?

Thanks!

**Comment by mikehardy** [comment, marked as answer]
There are no current plans. I can say, having watched the progress of this as a member of the release-testers group - and seeing the pain it causes in for example the reanimated module, that it looks quite expensive to me, on the "module maintainership" side. So that's a negative: very high cost in new tech integration as it brings in new toolchains and requires module release cadence to sync with react-native release cadence - a significant new hurdle for all library consumers and a significant new dependency for release cycles.

~~If I understand correctly the sole benefit is that it is a performance optimization~~ [Edit: this explains the benefits better, I'm still not sure they *really* apply to react-native-firebase, but it is important to understand them:https://github.com/react-native-community/discussions-and-proposals/discussions/546#discussioncomment-4160151  ]

In our case, I'm not sure there is any actual benefit to it for performance improvements with the exception of firestore, which could benefit perhaps from better performance in data marshalling/unmarshalling. But then it is my understanding that most of the performance issues people have had with firestore are related to local cached data being accessed slowly due to underlying issues in the native SDKs.

So to me I see large costs + a lot of added complexity and I see very uncertain benefits. Thus I haven't put it on my short list of things todo personally

Side note is that our testing infrastructure requires the JSI interface and I'm not sure that works when Fabric is enabled. So if I understand correctly, this is blocked until our e2e infrastructure moves to a pure on-device solution - that's something we have in progress but it's not done yet.

**Reply by mfbx9da4** [reply]
On the latter point, it is my understanding that turbo modules, fabric and the JSI are totally orthogonal pieces of the new architecture so enabling one shouldn't disrupt another.

**Reply by mikehardy** [reply]
I'll check in to that, I was under the impression that JSI could not be used if Fabric was enabled, but I could be wrong...

**Reply by dmahajan980** [reply]
Hello @mikehardy! It would be great if you can shed some light on supporting new architecture for this library. Although I am not an Android/iOS dev, I can try helping to the best of my knowledge to enable support for new architecture.

**Reply by mikehardy** [reply]
Nothing at the moment, no. PRs always welcome

**Reply by mr-ryan-james** [reply]
At some point, not supporting it is going to be a pretty visible shortcoming, as libraries supporting it are going to reach a important critical mass sometime in the future
https://github.com/reactwg/react-native-new-architecture/discussions/6

**Reply by mikehardy** [reply]
@mr-ryan-james that sounds great, and I welcome that future as soon as new architecture actually supports use_frameworks.
Right now, since the new architecture does not support use_frameworks at all, and we require use_frameworks, and Meta folks are on record as saying this is something they need to do internally because it is structural so does not lend itself to PRs, I'm not really sure what do about it.

What do you suggest?

**Reply by mikehardy** [reply]
(also possibly interesting to note that among the modules that are new-architecture-enabled, they are ingesting significant build system churn every single react-native release, including the upcoming 0.71 one - fundamental "how should we integrate and build this?" things that I'm happy to let them fuzz-bust through while we manage other cross-cutting changes internally, like #6705 ). All to say, I'm watching this space closely, but right now... :shrug: not sure there's a next step. Need Meta to unblock use_frameworks + new architecture first

**Reply by mr-ryan-james** [reply]
Indeed, looking at this tweet, it does look like there is a lot of active reworking being done every time a React Native version is released.

https://twitter.com/mrousavy/status/1590673049816162304

**Reply by mr-ryan-james** [reply]
I see your comment was listed as a line item in this discussion! 👏 
https://github.com/react-native-community/discussions-and-proposals/discussions/546#discussioncomment-4168642

Hopefully there is some traction to address the `use_frameworks` issue that you brought up there

**Reply by mikehardy** [reply]
Yep - I'm sure it will happen eventually, I'm not combative at all with the react-native core team rather I collaborate with them *all the time* to make sure releases work and there are no regressions (c.f. https://github.com/reactwg/react-native-releases/discussions/41#discussioncomment-4170555 / https://github.com/reactwg/react-native-releases/discussions/41#discussioncomment-4242750 for upcoming react-native 0.71)

Meta had its own corporate upheaval so I'm not sure where that work will find a home but I'm sure eventually new architecture and flipper will get use_frameworks working, and at the very least old architecture won't be removed until it does

**Reply by jphilipstevens** [reply]
Looks like someone is trying to help this cause with a PR: https://github.com/invertase/react-native-firebase/pull/6966

Ive started the migration to the new architecture and since I use analytics and dynamic links for a personal project i would love to see this make some progress

**Reply by affansk** [reply]
Finally, for iOS, the New Architecture is now compatible with the commonly used use_frameworks! CocoaPod configuration

**Reply by dmahajan980** [reply]
@affansk any links to PR/docs where we can read up on this?

**Reply by affansk** [reply]
@dmahajan980 i may be wrong. i am not able to find exact merged pr related to it.  but i just read on https://blixtdev.com/react-native-0-72-whats-new/

update : please check this : https://github.com/reactwg/react-native-new-architecture/discussions/115

**Reply by mikehardy** [reply]
That's fantastic! I have on my board to run new architecture compile tasks after seeing they were working on it for 0.72, I'll see if there are any gotchas, but in general this unblocks things here, fantastic news

**Reply by wkwyatt** [reply]
@mikehardy have there been any updates on this?

**Reply by mikehardy** [reply]
No updates at the moment, but we're quite active here in the repo. We're nearly done with the work to adopt the firebase-js-sdk v9 apis which was a higher priority but also a lot of work as it is an ecosystem shift similar to new architecture

Eventually we will have no choice but to adopt new architecture here of course, but the work has not really begun yet

**Reply by wkwyatt** [reply]
Is Firebase v9 natively compatible with the new architecture on RN?  I am running into issues loading Firebase through cocoa pods as well as using the react-native-firebase lib.  I know you mentioned this,
https://github.com/invertase/react-native-firebase/issues/6368#issuecomment-1186574446

So I wasn't sure if that was just for rn-firebase or for the Firebase cocoapod as well

**Comment by Nantris** [comment]
Any change on this? With new architecture moving closer to becoming the defacto standard, is it still not yet planned?

**Reply by bglgwyng** [reply]
https://github.com/invertase/react-native-firebase/pull/7688
It seems that New Architecture is now supported!

**Reply by TowhidKashem** [reply]
@bglgwyng can you please confirm it is? According to the main expo docs:

https://docs.expo.dev/guides/new-architecture/#known-issues-in-third-party-libraries

it's not :(

**Reply by bglgwyng** [reply]
I tested upgrading my app to new architecture but haven't tested all the firebase features yet. I think you'd better try new architecture + firebase yourself.

**Comment by BasixKOR** [comment]
I'm assmuming this has been solved by #7688, is it correct?

**Comment by younes0** [comment]
New architecture compatibility was announced by @brentvatne on this tweet https://x.com/notbrent/status/1832108444578672709
Thank you @russellwheatley & invertase team

**Comment by akshay-engage** [comment]
I do not see codegen configuration added to the firebase and so does the new architecture implementation. How is it new architecture supported? It is just working interoperatiblity layer only (which support libraries of old architecture in the new architecture app as well)

**Reply by russellwheatley** [reply]
React Native Firebase is currently going through a TypeScript migration. Once this is completed, we will move packages to using TurboModules. Cloud functions package is the only package that has moved to TurboModules at the moment. React Native Firebase is compatible with the latest architecture (i.e. latter versions of React Native), which is what has been announced. Therefore, you can use latter versions of React Native and React Native Firebase will work.
